### PR TITLE
[runtime] Throw a managed exception instead of trying to call a null function pointer if the runtime tries to call a function that has been linked away.

### PR DESCRIPTION
**Pull request by rolfbjarne**

### CODE REVIEW

File: runtime/delegates.inc.t4 [hunk 1]
```diff
@@ -27,10 +27,24 @@ struct Delegates {
...
+
 <# foreach (var d in delegates) { #>
 <#= d.CReturnType #>
 <#= d.EntryPoint #> (<#= d.CArgumentSignature #>)
 {
+	<#if (d.ExceptionHandling && d.OnlyDynamicUsage) {#>if (delegates.<#= d.EntryPoint.Substring ("xamarin_".Length) #> == NULL) {
+		*exception_gchandle = create_linked_away_exception ("<#= d.EntryPoint.Substring ("xamarin_".Length) #>");
+		return<# if (d.CReturnType != "void") { #> (<#= d.CReturnType #>) 0<# } #>;
+	}
+<#}#>
```
Comment (rolfbjarne) [inline runtime/delegates.inc.t4]: The T4 code is somewhat hard to read, but the generated function now look like this:

```c
MonoObject *
xamarin_get_block_wrapper_creator (MonoObject * method, int parameter, guint32 *exception_gchandle)
{
	if (delegates.get_block_wrapper_creator == NULL) {
		*exception_gchandle = create_linked_away_exception ("get_block_wrapper_creator");
		return (MonoObject *) 0;
	}
	return delegates.get_block_wrapper_creator (method, parameter, exception_gchandle);
}
```

The `if (delegates.get_block_wrapper_creator == NULL) { ... }` block is the new code.

File: docs/website/mmp-errors.md [hunk 2]
```diff
@@ -522,3 +522,9 @@ There are a few reasons this may happen:
 
 * It could be a bug in Xamarin.Mac. If this is the case, please file a bug at
   [https://bugzilla.xamarin.com](https://bugzilla.xamarin.com/enter_bug.cgi?product=Xamarin.Mac).
+
+### <a name="MM8028"/>MM8028: The runtime function {function} has been linked away.
+
+This usually indicates a bug in Xamarin.Mac, because runtime functions should
+not be linked away if they're needed. Please file a bug at
+[https://bugzilla.xamarin.com](https://bugzilla.xamarin.com/enter_bug.cgi?product=Xamarin.Mac).
```
Comment (spouliot) [inline docs/website/mmp-errors.md]: start using the GitHub link for new errors
we'll do a larger round once (xcode10) is merged back into master (to minimize the conflicts)

File: runtime/xamarin/runtime.h [hunk 3]
```diff
@@ -233,6 +233,7 @@ void			xamarin_process_nsexception_using_mode (NSException *ns_exception, bool t
 void			xamarin_process_managed_exception (MonoObject *exc);
 void			xamarin_process_managed_exception_gchandle (guint32 gchandle);
 void			xamarin_throw_product_exception (int code, const char *message);
+guint32			xamarin_create_product_exception (int code, const char *message);
```
Comment (spouliot) [inline runtime/xamarin/runtime.h]: minor: mixed tabs/spaces ?

Comment (rolfbjarne) [reply runtime/xamarin/runtime.h]: No, just 4-vs-8 spaces for tabs (the file is indented to match 4 space, GitHub renders using 8 spaces).

You can see the same thing happening elsewhere in the file if you expand the diff.

**Pull request merged**